### PR TITLE
NaN can be used as a key in Map

### LINKED ISSUE #5833: NaN cannot be used as a key in Map
The primary motivation for`Map` existance is implementation of `Vector.distinct`. To make such implementation easy/fast/possible we **don't want errors** from `Map.insert`. On contrary, we want **every object** to provide **hash code** - so it can be used as a key for the `Map`. Constraints:

1. it is desirable for Map to always satisfy: `m.insert k v . contains_key k == True`
2. For `Number` we are going to use JavaScript's [Equality comparisons and sameness](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Equality_comparisons_and_sameness#same-value-zero_equality) to ensure `NaN` works as a key
3. Custom types - shall have a **forced reflexivity** - e.g. `Meta.is_same_object` implies that they are `Any.==` which in turn satisfies the first point

Following code shows the expected behavior for custom types that violate reflexivity:
```haskell
type My_Nan
    Value comment:Text

...comparator such that it always returns Nothing

k = My_Nan.Value "foo"
k2 = My_Nan.Value "foo"
(k==k).should_be_true
(k==k2).should_be_false
m = Map.empty.insert k 10
m.contains_key k . should_be_true
m.get k . should_equal 10
m.contains_key k2 . should_be_false

m2 = m.insert k2 20
m2.get k . should_equal 10
m2.get k2 . should_equal 20
m2.size . should_equal 2

m3 = m2.insert k 30
m3.size . should_equal 2
m3.get k . should_equal 30
```

e.g. , we will do a best effort solution. Thanks to `Any.==` relying on `Meta.is_same_object`, even if `compare` implementation is **always incomparable**, `x==x` will be `True`. Multiple separate instances of incomparable atoms will however not be. 

Follow up of 
- #5771 
- #5818

```[tasklist]
### Tasks
- [ ] Make sure `NaN` can be a key in `Map` using the JavaScript's [Equality comparisons and sameness](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Equality_comparisons_and_sameness#same-value-zero_equality)
- [x] Make sure `x == x` via `Meta.is_same_object` (tracked in #6065)
- [ ] Update tests
- [ ] Update documentation
```

### LINKED ISSUE #6065: Any.== should check for Meta.is_same_object as the first step
Mentioned in https://github.com/enso-org/enso/issues/5742#issuecomment-1444995335, `Meta.is_same_object obj1 obj2` should imply `obj1 == obj2`. So far, the reason why it is not checked as the first step is because of performance - explained in https://github.com/enso-org/enso/issues/5709#issuecomment-1463609295.

When applying this patch:
```diff
diff --git a/distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Day_Of_Week.enso b/distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Day_Of_Week.enso
index 4e3dd7400..5aca27d98 100644
--- a/distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Day_Of_Week.enso
+++ b/distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Day_Of_Week.enso
@@ -1,4 +1,5 @@
 import project.Data.Numbers.Integer
+from project.Data.Ordering import all
 from project.Data.Boolean import Boolean, True, False
 
 polyglot java import java.time.DayOfWeek
@@ -49,3 +50,14 @@ type Day_Of_Week
         Day_Of_Week.Thursday -> DayOfWeek.THURSDAY
         Day_Of_Week.Friday -> DayOfWeek.FRIDAY
         Day_Of_Week.Saturday -> DayOfWeek.SATURDAY
+
+## PRIVATE
+type Day_Of_Week_Comparator
+    compare x y =
+        hash_x = Day_Of_Week_Comparator.hash x
+        hash_y = Day_Of_Week_Comparator.hash y
+        Comparable.from hash_x . compare hash_x hash_y
+
+    hash x = x.to_integer
+
+Comparable.from (_:Day_Of_Week) = Day_Of_Week_Comparator
```

`Day_Of_Week.Monday < Day_Of_Week.Friday` will cause StackOverflow because of that.

### PR DESCRIPTION
**Pull request by Akirathan**
Fixes #5833 #6065 

### Pull Request Description

`Number.nan` can be used as a key in `Map`. This PR basically implements the support for [JavaScript's Same Value Zero Equality](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Equality_comparisons_and_sameness#same-value-zero_equality) so that `Number.nan` can be used as a key in `Map`.

### Important Notes
- For NaN, it holds that `Meta.is_same_object Number.nan Number.nan`, and `Number.nan != Number.nan` - inspired by JS spec.
- `Meta.is_same_object x y` implies `Any.== x y`, except for `Number.nan`.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

**Comment by mwu-tow** [comment]
@Akirathan 
> `Number.nan` can be used as a key in `Map`.

Why do we want this? For me `NaN` is the value that is not equal to itself. That by definition kind of prevents it from being useful as a key in the map.
Even if we wanted this, I'd rather expect some kind of byte-representation-comparison rather than just `NaN==NaN` assumption.  There are different values (byte-wise) that are considered to be a 'NaN'. Do we want to disregard the differences?

**Comment by Akirathan** [comment]
> @Akirathan
> 
> > `Number.nan` can be used as a key in `Map`.
> 
> Why do we want this? For me `NaN` is the value that is not equal to itself. That by definition kind of prevents it from being useful as a key in the map. Even if we wanted this, I'd rather expect some kind of byte-representation-comparison rather than just `NaN==NaN` assumption. There are different values (byte-wise) that are considered to be a 'NaN'. Do we want to disregard the differences?

As noted in the issue description #5833, we want to comply with [JavaScript's same-value-zero equality](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Equality_comparisons_and_sameness#same-value-zero_equality). The reason for that is because we want *every* value in Enso, including polyglot values, to have some hash code, and thus be used as a key in the map.

File: test/Tests/src/Data/Map_Spec.enso [hunk 1]
```diff
@@ -143,6 +168,33 @@ spec =
             (map.get key_map).should_equal 23
             (map.get map).should_equal Nothing
 
+        Test.specify "should support another hash map with NaN keys as key" <|
+            Map.singleton (Map.singleton Number.nan 1) 42 . size . should_equal 1
+            Map.singleton (Map.singleton Number.nan 1) 42 . keys . at 0 . keys . to_text . should_equal "[NaN]"
+            Map.singleton (Map.singleton Number.nan 1) 42 . keys . at 0 . get Number.nan . should_equal 1
+            Map.singleton (Map.singleton Number.nan 1) 42 . at (Map.singleton Number.nan 1) . should_equal 42
+
+        Test.specify "should support atoms with custom comparators that violate reflexivity as keys" <|
```
Comment (Akirathan) [inline test/Tests/src/Data/Map_Spec.enso:177]: This test was copied from the issue description

File: test/Tests/src/Semantic/Meta_Spec.enso [hunk 2]
```diff
@@ -281,13 +281,18 @@ spec =
...
-    Test.group "Check Nothing" <|
+    Test.group "Check Nothing and NaN" <|
         Test.specify "Nothing.is_a Nothing" <|
             Nothing.is_a Nothing . should_be_true
+            Meta.is_same_object Nothing Nothing . should_be_true
 
         Test.specify "type_of Nothing is Nothing" <|
             Meta.type_of Nothing . should_equal Nothing
 
+        Test.specify "NaN and NaN should be the same object" <|
```
Comment (Akirathan) [inline test/Tests/src/Semantic/Meta_Spec.enso:292]: This is exactly how JS works, according to the spec in [Same value zero equality](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Equality_comparisons_and_sameness#same-value-zero_equality)

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Map.enso [hunk 3]
```diff
@@ -15,6 +15,12 @@ from project import Error, Nothing, Any, Panic
    for equality any objects that can appear in Enso - primitives, Atoms, values coming from
    different languages, etc.
 
+   For keys that are not reflexive, like `Number.nan`, "Same Value Equality", from JavaScript
+   specification in https://developer.mozilla.org/en-US/docs/Web/JavaScript/Equality_comparisons_and_sameness#same-value-zero_equality,
```
Comment (JaroslavTulach) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Map.enso]: I think we can use Markdown format in the comments. Then I recommend to use
```markdown
[Same Value Equality](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Equality_comparisons_and_sameness#same-value-zero_equality)
```
e.g. hyperlink text without the URL being visible.

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java [hunk 4]
```diff
@@ -214,43 +209,43 @@ boolean equalsStrings(Object selfString, Object otherString,
...
 
   @Specialization
-  boolean equalsAtoms(Atom self, Atom other, @Cached EqualsAtomNode equalsNode) {
-    return equalsNode.execute(self, other);
+  boolean equalsAtoms(
+      Atom self,
+      Atom other,
+      @Cached EqualsAtomNode equalsNode,
+      @Cached IsSameObjectNode isSameObjectNode) {
+    return isSameObjectNode.execute(self, other) || equalsNode.execute(self, other);
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java:232]: This is the fix for #6065, right?

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/IsSameObjectNode.java [hunk 5]
```diff
@@ -23,6 +25,15 @@ public static IsSameObjectNode build() {
 
   public abstract boolean execute(@AcceptsError Object left, @AcceptsError Object right);
 
+  @Specialization
+  boolean isSameDouble(double left, double right) {
+    if (Double.isNaN(left) && Double.isNaN(right)) {
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/IsSameObjectNode.java:30]: This is the handling of `NaN` as being _"same"_.

File: test/Tests/src/Semantic/Meta_Spec.enso [hunk 6]
```diff
@@ -281,13 +281,18 @@ spec =
...
+            Meta.is_same_object Number.nan Number.nan . should_be_true
+            (Number.nan == Number.nan) . should_be_false
```
Comment (JaroslavTulach) [inline test/Tests/src/Semantic/Meta_Spec.enso:294]: `nan != nan`, but `Meta.is_same_object` gives `true`. OK.

**Review by JaroslavTulach — APPROVED** [review]

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java (hunk 4 above)
Comment (Akirathan) [reply engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java:232]: Yes, #6065 is addressed in these commits:
- 35074849afa320cac15d866b9c7f3921ef3f2bc4
- 55ceaba9b1ff2a0cf63fec5b11aabbad834c0e06

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/IsSameObjectNode.java [hunk 7]
```diff
@@ -23,6 +25,15 @@ public static IsSameObjectNode build() {
...
+      return true;
+    } else {
+      return left == right;
+    }
```
Comment (hubertp) [inline engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/IsSameObjectNode.java:34]:
```suggestion
    return (Double.isNaN(left) && Double.isNaN(right)) || (left == right);
```
personal preference feel free to ignore :p

File: tools/performance/engine-benchmarks/bench_download.py [hunk 8]
```diff
@@ -706,14 +706,20 @@ def commit_to_str(commit: Commit) -> str:
...
+            print(
+                f"No successful benchmarks found within period since {since} until {until}")
             exit(1)
         job_reports: List[JobReport] = []
         for bench_run in bench_runs:
             job_report = get_bench_report(bench_run, cache, temp_dir)
             if job_report:
                 job_reports.append(job_report)
         logging.debug(f"Got {len(job_reports)} job reports")
+        if len(job_reports) == 0:
```
Comment (hubertp) [inline tools/performance/engine-benchmarks/bench_download.py:718]: That seems unrelated?

**Review by hubertp — APPROVED** [review]
Do we want/need any tests for polyglot handling of NaNs?

File: tools/performance/engine-benchmarks/bench_download.py (hunk 8 above)
Comment (Akirathan) [reply tools/performance/engine-benchmarks/bench_download.py:718]: Yes. I added it because just yesterday I uploaded new benchmark results, and stepped into this issue, so I committed this tiny change along in this PR.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Map.enso (hunk 3 above)
Comment (Akirathan) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Map.enso]: Done in fc5aace5a9cd0cc3a3a00ba30dfd2fcb1199b96e

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java [hunk 9]
```diff
@@ -32,9 +33,9 @@
       Does not throw dataflow errors or panics.
 
       Note that this is different than `Meta.is_same_object`, which checks whether two
-      references point to the same object on the heap.
-      """
-)
+      references point to the same object on the heap. However `Any.==` implies
+      `Meta.is_same_object` for all object with the exception of `Number.nan`.
+      """)
```
Comment (radeusgd) [inline engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java]: This does not seem true, or am I misunderstanding?

If I have two separate instances of an object, they may be `==` but will not be `is_same_object`, right?

Shouldn't this condition be contraposed?

If I have `is_same_object x y` then I'm guranteed that `x == y`, as long as they are not `NaN`.

Comment (Akirathan) [reply engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java]: You are right. I wrote the implication in the wrong direction. Fixed in d65e5e9f38a7b8b77e9673447af301744ff2d0a7.

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java [hunk 10]
```diff
@@ -44,11 +45,7 @@ public static EqualsNode build() {
 
   public abstract boolean execute(@AcceptsError Object self, @AcceptsError Object right);
 
-  /**
-   * Primitive values
-   **/
-
-
+  /** Primitive values */
```
Comment (radeusgd) [inline engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java]: This used to be a 'section' comment, but now reformatting seems to have attached it to the `equalsBoolBool` method.

That doesn't seem right.

I think we should either keep these as 'sections' or remove them. Being attached to the first method is just awkward.

File: test/Tests/src/Data/Vector_Spec.enso [hunk 11]
```diff
@@ -654,8 +654,23 @@ type_spec name alter = Test.group name <|
...
+        alter [1, "a"] . distinct . should_equal [1, "a"]
+        alter ["a", 1] . distinct . should_equal ["a", 1]
+        alter [Nothing, Nothing] . distinct . should_equal [Nothing]
+        alter [Number.nan, Number.nan] . distinct . to_text . should_equal "[NaN]"
+        alter [Nothing, Number.nan, Nothing, Number.nan] . distinct . to_text . should_equal "[Nothing, NaN]"
+        alter [1, Nothing] . distinct . should_equal [1, Nothing]
+        alter [Nothing, 1, Nothing] . distinct . should_equal [Nothing, 1]
+        alter [1, Number.nan] . distinct . to_text . should_equal "[1, NaN]"
+        alter [Number.nan, 1, Number.nan] . distinct . to_text . should_equal "[NaN, 1]"
+        alter [1, Number.nan, 1] . distinct . to_text . should_equal "[1, NaN]"
```
Comment (radeusgd) [inline test/Tests/src/Data/Vector_Spec.enso:677]: How does `distinct` work with other non-reflexive values, like `My_Nan`? Can we get a test for it too?

**Review by radeusgd — APPROVED** [review]
Looks good, I'd just appreciate a test for `My_Nan` in `distinct`.

This also should at some point work consistently in `Table.aggregate` - `Aggregate_Column.Group_By`. I guess no need to add these tests now, but maybe we should add a note to #6292 to not forget about it? It is quite a brittle feature and easy to get wrong so it will be good to have tests for this case once available.

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java (hunk 10 above)
Comment (Akirathan) [reply engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java]: You are right. I will remove it. Recently, I enabled a plugin in my IDE that does automatic reformatting on commits and this is the result.

File: test/Tests/src/Data/Vector_Spec.enso (hunk 11 above)
Comment (Akirathan) [reply test/Tests/src/Data/Vector_Spec.enso:677]: Added in 6bae7283fe1ce8276de282dcee37ef45abaa7d89
